File: src/Adaptive.ReactiveTrader.Client.Domain/ServiceClients/BlotterServiceClient.cs
using System;
using System.Collections.Generic;
using System.Reactive;
using System.Reactive.Disposables;
using System.Reactive.Linq;
using Adaptive.ReactiveTrader.Client.Domain.Transport;
using Adaptive.ReactiveTrader.Shared;
using Adaptive.ReactiveTrader.Shared.DTO.Execution;
using Adaptive.ReactiveTrader.Shared.Logging;
using Microsoft.AspNet.SignalR.Client;

namespace Adaptive.ReactiveTrader.Client.Domain.ServiceClients
{
    internal class BlotterServiceClient : ServiceClientBase, IBlotterServiceClient
    {
        private readonly ILog _log;

        public BlotterServiceClient(IConnectionProvider connectionProvider, ILoggerFactory loggerFactory) : base(connectionProvider)
        {
            _log = loggerFactory.Create(typeof (BlotterServiceClient));
        }

        public IObservable<IEnumerable<TradeDto>> GetTradesStream()
        {
            return GetResilientStream(connection => GetTradesForConnection(connection.BlotterHubProxy), TimeSpan.FromSeconds(5));
        }

        private IObservable<IEnumerable<TradeDto>> GetTradesForConnection(IHubProxy blotterHubProxy)
        {
            return Observable.Create<IEnumerable<TradeDto>>(observer =>
            {
                // subscribe to trade feed first, otherwise there is a race condition 
                var spotTradeSubscription = blotterHubProxy.On<IEnumerable<TradeDto>>(ServiceConstants.Client.OnNewTrade, observer.OnNext);

                _log.Info("Sending blotter subscription...");
                var sendSubscriptionDisposable = SendSubscription(blotterHubProxy)
                    .Subscribe(
                        _ => _log.InfoFormat("Subscribed to blotter."),
                        observer.OnError);

                var unsubscriptionDisposable = Disposable.Create(() =>
                {
                    // send unsubscription when the observable gets disposed
                    _log.Info("Sending trades unsubscription...");
                    SendUnsubscription(blotterHubProxy)
                        .Subscribe(
                            _ => _log.InfoFormat("Unsubscribed from blotter."),
                            ex => _log.WarnFormat("An error occurred while unsubscribing from blotter: {0}", ex.Message));
                });
                return new CompositeDisposable { spotTradeSubscription, unsubscriptionDisposable, sendSubscriptionDisposable };
            })
                .Publish()
                .RefCount();
        }

        private static IObservable<Unit> SendSubscription(IHubProxy blotterHubProxy)
        {
            return Observable.FromAsync(() => blotterHubProxy.Invoke(ServiceConstants.Server.SubscribeTrades));
        }

        private static IObservable<Unit> SendUnsubscription(IHubProxy blotterHubProxy)
        {
            return Observable.FromAsync(() => blotterHubProxy.Invoke(ServiceConstants.Server.UnsubscribeTrades));
        } 
    }
}

File: src/Adaptive.ReactiveTrader.Web/Client/ServiceClients/BlotterServiceClient.ts
 class BlotterServiceClient extends ServiceClientBase implements IBlotterServiceClient {
     constructor(connectionProvider: IConnectionProvider) {
         super(connectionProvider);
     }

     getTradesStream(): Rx.Observable<TradeDto[]> {
         return super.getResilientStream(connection=> this.getTradesStreamFromConnection(connection), 5000);
     }

     private getTradesStreamFromConnection(connection: IConnection): Rx.Observable<TradeDto[]> {
         return Rx.Observable.create<TradeDto[]>(observer=> {
             var tradesSubscription = connection.allTrades.subscribe(observer);

             console.log("Sending blotter subscription...");
             connection.blotterHubProxy.invoke("SubscribeTrades")
                 .done(_=> console.log("Subscribed to blotter"))
                 .fail(ex=> observer.onError(ex));

             var unsubscriptionDisposable = Rx.Disposable.create(()=> {
                 console.log("Sending blotter unsubscription...");

                 connection.blotterHubProxy.invoke("UnsubscribeTrades")
                     .done(_=> console.log("Unsubscribed from blotter stream."))
                     .fail(ex=> console.error("An error occured while unsubscribing from blotter:" + ex));
             });

             return new Rx.CompositeDisposable([tradesSubscription, unsubscriptionDisposable]);
         })
         .publish()
         .refCount();
     }
 }